ddr/lib/ddr-ir/Macro.cpp

```diff
@@ -25,57 +25,951 @@
 #include "ddr/ir/Macro.hpp"
 
 #include "ddr/std/sstream.hpp"
+#include <ctype.h>
 #include <string.h>
+#include <stdint.h>
+#include <algorithm>
+#include <vector>
 
 using std::string;
 using std::stringstream;
 
-DDR_RC
-Macro::getNumeric(long long *ret)
+enum Symbol {
+	OR_OR,		/* || */
+	AND_AND,	/* && */
+	OR,			/* | */
+	XOR,		/* ^ */
+	AND,		/* & */
+	EQ_EQ,		/* == */
+	NOT_EQ,		/* != */
+	GT,			/* > */
+	LT,			/* < */
+	GT_EQ,		/* >= */
+	LT_EQ,		/* <= */
+	GT_GT,		/* >> */
+	LT_LT,		/* << */
+	PLUS,		/* + */
+	MINUS,		/* - */
+	MULT,		/* * */
+	DIV,		/* / */
+	REM,		/* % */
+	NOT,		/* ! */
+	TILDE,		/* ~ */
+	LPAREN,		/* ( */
+	RPAREN		/* ) */
+};
+
+class MacroScanner
 {
-	DDR_RC rc = DDR_RC_OK;
-	string value = getValue();
+private:
+	char const *_cursor;
+	char const *_end;
 
-	if (value.length() > 0) {
-		/* If the pre-processed macro contains no brackets, read it as a number. */
-		if (string::npos == value.find('(')) {
-			stringstream ss;
-			ss << value;
-			long long valueNumeric = 0;
-			ss >> valueNumeric;
-			if (ss.fail()) {
-				rc = DDR_RC_ERROR;
-			} else if (NULL != ret) {
-				*ret = valueNumeric;
-			}
+	MacroScanner(string expression) :
+		_cursor(expression.c_str()),
+		_end(_cursor + expression.length())
+	{
+	}
+
+	bool atSymbol(Symbol sym);
+	bool readNumber(int64_t *ret);
+
+	friend class MacroParser;
```

Comment:
Make the required methods public rather than use a friend declaration.

FYI, there should be a forward declaration for `class MacroParser` before it can be meaningfully referenced in a friend declaration.